### ISSUE TITLE
Nan question in SupConLoss

### ISSUE THREAD

**Issue by kkj9333** [emadeldeen24/TS-TCC#25]
I found that in 'SupConLoss::forward' of 'loss.py', logits are defined by anchor_Dot_Contrast subtracts the maximum value of each line from itself, but if the distribution of anchor_Dot_Contrast is uneven, where the maximum value of a certain row is much larger than the other values of that row, will result in the logits being composed of some smaller negative numbers (such as -208.3) and 0. Due to the use of torch. float32 precision in the code, in this case, there is only one 1 in a certain line of torch. exp (logits), and the rest are all zero (and approximately zero); Even worse, if it's related to logits_ Multiplying masks will result in exp_ All logits in a certain line are 0. This will result in a log_ Prob=logits - torch. log (exp_logits. sum (1, keepdim=True)) becomes infinite, resulting in the loss being calculated as Nan. This will be an unacceptable consequence.

**Comment by emadeldeen24** [comment]
Hi,
thanks for reporting this. This issue is more related to the [CA-TCC work](https://github.com/emadeldeen24/CA-TCC), where we use the `SupConLoss`. 

To summarize what I understood:
you mean that the process of subtracting the maximum value from each row in the `anchor_Dot_Contrast` tensor can lead to some extremely negative values in the `logits` tensor, especially when there's a large disparity between the maximum value and other values in a particular row.
As a consequence, the exponential operation `(torch.exp(logits))` and `(torch.log(exp_logits.sum(1, keepdim=True)))`, can result in numerical instability, leading to values of NaN for the loss.

I think you are right, and this can be a valid point. We can address it in the implementation in several ways. For example, we can do clamping to the values of logits before performing the exponential operation. This can be to a predefined range.
Another way is to add a small constant before taking the logarithm to prevent the logarithm of a value close to zero. 

It would be very beneficial if you can post the whole error message so that I can check and debug it properly.

**Comment by kkj9333** [comment]
Yes, that's what I mean. During the training process, the loss becomes Nan, resulting in all weights becoming Nan, which will not lead to python errors. 
![image](https://github.com/emadeldeen24/TS-TCC/assets/51207072/99f0f213-99b4-494d-9e15-f302018c45d0)
But I can provide information on debugging SupConLoss:
- When entering 'forward':
![image](https://github.com/emadeldeen24/TS-TCC/assets/51207072/c7e0af16-8e13-472a-a844-eb7edeb7053e)
- Before subtracting the maximum value:
![image](https://github.com/emadeldeen24/TS-TCC/assets/51207072/ae5fbe8a-87c2-4f2e-9dd0-7721c03ce52e)
- After subtracting the maximum value:
![image](https://github.com/emadeldeen24/TS-TCC/assets/51207072/84e1e569-c39d-4713-b1d0-0936e343a75f)
- `exp_logits = torch.exp(logits) * logits_mask` After exponential operation, it will become zero.
![image](https://github.com/emadeldeen24/TS-TCC/assets/51207072/cc542c49-43f4-4c2a-9bc4-8c1e3e43cae1)
- When calculating `log_prob`, some calculation results become Nan due to `-torch. log` Nan appeared in the final calculation of the loss.:
![image](https://github.com/emadeldeen24/TS-TCC/assets/51207072/5d9adaaf-1f37-49fe-99c9-24f9a5922c74)

 My current solution is to add` eps=1e-15` when calculating the row sum of `exp_logits`to avoid zero. But will this lead to situations where losses cannot converge?

**Comment by kkj9333** [comment]
If the `eps` I set is too small, this error message will still appear:
![image](https://github.com/emadeldeen24/TS-TCC/assets/51207072/844c371e-0952-4e40-82ee-306879517932)

**Comment by emadeldeen24** [comment]
How about clamping logits to a predefined range, say, [-1000, 100]? `torch.clamp(logits, min=-100, max=100)`.

**Comment by kkj9333** [comment]
Thank you for your help! Perhaps I can use some methods to alleviate this problem myself.
